source/common/modules/markdown-editor/commands/transforms/transform-selected-text.ts

```diff
@@ -0,0 +1,72 @@
+/**
+ * @ignore
+ * BEGIN HEADER
+ *
+ * Contains:        transformSelectedText function
+ * CVM-Role:        Utility function
+ * Maintainers:     Rich Douglas
+ * License:         GNU GPL v3
+ *
+ * Description:     A builder of `Commands` for transforming selected text.
+ *
+ * END HEADER
+ */
+
+import { ChangeSet, type StateCommand } from '@codemirror/state'
+
+/**
+ * Transform the supplied `text`.
+ *
+ * * strip duplicate spaces
+ * * trim leading and trailing whitespace
+ * * change the casing
+ * * etc.
+ *
+ * @param   {string}  text  The text to be transformed.
+ *
+ * @return  {string}        The transformed text.
+ */
+export type TransformText = (text: string) => string
+
+/**
+ * Return a `Command` to transform selected text.
+ *
+ * @param   {TransformText} transform The transformation to be applied to the
+ *                                    selected text.
+ *
+ * @return  {StateCommand}            A `Command` to transform selected text.
+ */
+export function transformSelectedText (transform: TransformText): StateCommand {

```

Comment:
This function encapsulates all the machinery of transforming selected text using the CodeMirror API.

Individual transforms are then plugged into this function -- via the `transform` parameter -- and don't need to duplicate logic and/or concern themselves with the CodeMirror API.

This is a `StateCommand` -- which just requires `state` and `dispatch` -- rather than a more general `Command` because it makes these transforms easier to (unit) test because we don't have to create a "full fat" `EditorView`.

* [StateCommand](https://codemirror.net/docs/ref/#state.StateCommand)